File: 0442-find-all-duplicates-in-an-array/0442-find-all-duplicates-in-an-array.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        vector<int> result;
        
        for (int i = 0; i < nums.size(); ++i) {
            int index = abs(nums[i]) - 1;
            
            // If the number at this index is negative, it means the number has been seen before
            if (nums[index] < 0) {
                result.push_back(index + 1);
            } else {
                // Mark the number as seen by negating the value at that index
                nums[index] = -nums[index];
            }
        }
        
        return result;
    }
};

File: 0442-find-all-duplicates-in-an-array/0442-find-all-duplicates-in-an-array.py
class Solution:
    def findDuplicates(self, nums):
        result = []

        # Iterate through each number in the array
        for num in nums:
            index = abs(num) - 1

            # Check if the value at this index has been negated (indicating a duplicate)
            if nums[index] < 0:
                result.append(index + 1)
            else:
                # Negate the value at this index to mark the number as seen
                nums[index] = -nums[index]

        return result